File: server/internal/worldfield/worldfield.go
// Package worldfield answers what the world *is* at a position: which danger
// band it sits in, which biome covers it, and what material grade the ground
// there yields. It holds no world state and reads no table — every input
// arrives as a plain Params value — so the simulation, the loader's coverage
// check, and the browser all reach one answer from one implementation.
//
// The browser's copy lives in web/src/game/worldfield.ts and must stay
// bit-identical to this one. Two rules keep that promise cheap:
//
//   - Every hash is 32-bit integer arithmetic, which JavaScript reproduces
//     exactly through Math.imul and the unsigned shift operators. A 64-bit
//     hash would need BigInt on the other side.
//   - Every float operation is +, -, *, /, or sqrt, all of which IEEE-754
//     specifies to the bit. math.Pow and JavaScript's Math.pow are *not*
//     required to agree, so the radial compression below is written as two
//     square roots rather than as an exponent, and the reward curve is
//     piecewise linear rather than a power.
package worldfield

import (
	"fmt"
	"math"
	"sort"
)

// Band is one radial danger ring, ordered outward from the hub.
type Band struct {
	ID, Name, PvP, MaterialGrade, Shape, Summary string
	Tier                                         int
	OuterRadius                                  float64
}

// Protected reports whether PvP damage is refused inside the band.
func (b Band) Protected() bool { return b.PvP == "off" || b.PvP == "restricted" }

// Safe reports whether the band offers the full service set — crafting,
// loadout mutation, respawn.
func (b Band) Safe() bool { return b.PvP == "off" }

// GradePoint is one vertex of the reward curve: at `At` of the world radius the
// ground is worth `Richness`. Between vertices the curve is linear, and the
// vertices are required to be convex, which is what makes the middle bands a
// route rather than the best farm.
type GradePoint struct{ At, Richness float64 }

// GradeThreshold is the richness a material grade starts at.
type GradeThreshold struct {
	ID   string
	Tier int
	At   float64
}

type GradeCurve struct {
	Points     []GradePoint
	Thresholds []GradeThreshold
}

// Params is everything the field is derived from. It is deliberately primitive:
// tuning builds one of these, and so does the client, so neither the field nor
// its coverage check depends on the table types.
type Params struct {
	Radius float64
	Seed   uint32
	// RegionCell is the biome lattice edge, measured in the compressed space
	// below rather than in world units.
	RegionCell float64
	// RegionJitter is how far a region site may wander inside its own cell, as
	// a fraction of the cell. Bounded below 1 so the five-cell search around a
	// query point always contains the nearest site.
	RegionJitter float64
	// RadialReference is the radius the compression leaves untouched. Positions
	// are pulled toward it by a fixed 3/4 power before the lattice is sampled,
	// so a biome is a broad radial swathe rather than a blob: every element
	// reaches every danger band, which is the two-axis rule biome-type and
	// radius-grade are built on.
	RadialReference float64
	// WarpCell and WarpAmplitude bend the lattice with value noise, so region
	// borders are ragged rather than straight Voronoi seams.
	WarpCell, WarpAmplitude float64
	// BlendWidth is how wide a border blend is, as a fraction of RegionCell.
	BlendWidth float64
	Biomes     []string
	Bands      []Band
	Grades     GradeCurve
}

// Biome is what covers a position: the region's biome, the nearest different
// one, and how far from their shared border the position sits. Blend is 0 on
// the border and 1 once the region is unambiguous, which is what a renderer
// cross-fades an ambient palette over.
type Biome struct {
	ID, Neighbour string
	Blend         float64
}

// Grade is what the ground is worth: the material grade it yields and the
// convex richness the yield itself scales with inside that grade.
type Grade struct {
	ID       string
	Tier     int
	Richness float64
}

// Region is the whole answer at a position, including a stable identity for
// the biome region itself so per-region parameters can be keyed on it.
type Region struct {
	ID     string
	Danger Band
	Biome  Biome
	Grade  Grade
}

type Field struct{ p Params }

// New copies the parameters and sorts the biome list, because a site's biome is
// drawn by index and the draw has to be stable against table ordering.
func New(p Params) *Field {
	field := &Field{p: p}
	field.p.Biomes = append([]string(nil), p.Biomes...)
	sort.Strings(field.p.Biomes)
	field.p.Bands = append([]Band(nil), p.Bands...)
	field.p.Grades.Points = append([]GradePoint(nil), p.Grades.Points...)
	field.p.Grades.Thresholds = append([]GradeThreshold(nil), p.Grades.Thresholds...)
	return field
}

func (f *Field) Params() Params { return f.p }

// DangerAt resolves the band containing a position. Anything past the rim
// resolves to the outermost band rather than to nothing, so a body pushed
// outside the world is never in an undefined zone.
func (f *Field) DangerAt(x, y float64) Band {
	distance := math.Sqrt(x*x + y*y)
	for _, band := range f.p.Bands {
		if distance <= band.OuterRadius {
			return band
		}
	}
	return f.p.Bands[len(f.p.Bands)-1]
}

// Protected reports whether PvP damage is refused at a position, and Safe
// whether the full service set is available there. Both exist so callers ask
// the field a question rather than comparing a radius of their own.
func (f *Field) Protected(x, y float64) bool { return f.DangerAt(x, y).Protected() }
func (f *Field) Safe(x, y float64) bool      { return f.DangerAt(x, y).Safe() }

// GradeAt walks the convex reward curve and reports the grade the richness has
// reached. The curve is what makes the rim disproportionately worth the walk:
// richness rises slowly across the Fringe and steeply toward the Deadlands, so
// the middle bands are a route rather than a destination.
func (f *Field) GradeAt(x, y float64) Grade {
	richness := f.RichnessAt(x, y)
	grade := Grade{Richness: richness}
	for _, threshold := range f.p.Grades.Thresholds {
		if richness >= threshold.At {
			grade.ID, grade.Tier = threshold.ID, threshold.Tier
		}
	}
	return grade
}

// RichnessAt is the reward curve itself, in [0,1].
func (f *Field) RichnessAt(x, y float64) float64 {
	points := f.p.Grades.Points
	if len(points) == 0 || f.p.Radius <= 0 {
		return 0
	}
	at := math.Sqrt(x*x+y*y) / f.p.Radius
	if at <= points[0].At {
		return points[0].Richness
	}
	for index := 1; index < len(points); index++ {
		previous, current := points[index-1], points[index]
		if at > current.At {
			continue
		}
		span := current.At - previous.At
		if span <= 0 {
			return current.Richness
		}
		return previous.Richness + (current.Richness-previous.Richness)*(at-previous.At)/span
	}
	return points[len(points)-1].Richness
}

// BiomeAt samples the noise-warped, radially compressed Voronoi lattice.
func (f *Field) BiomeAt(x, y float64) Biome {
	_, biome := f.regionAt(x, y)
	return biome
}

// RegionAt is the whole field at one position.
func (f *Field) RegionAt(x, y float64) Region {
	cell, biome := f.regionAt(x, y)
	return Region{
		ID:     fmt.Sprintf("region-%d:%d", cell.x, cell.y),
		Danger: f.DangerAt(x, y),
		Biome:  biome,
		Grade:  f.GradeAt(x, y),
	}
}

type cell struct{ x, y int64 }

// regionAt resolves the owning lattice site and the border blend around it.
//
// The search is five cells wide rather than three. A site may sit anywhere
// inside its own cell, so with the jitter bounded below one cell the nearest
// site to a point at a cell corner can still be two cells away on an axis; a
// three-cell search would occasionally pick the wrong region, and a region
// border that flickers is worse than one that costs twenty-five hashes.
func (f *Field) regionAt(x, y float64) (cell, Biome) {
	if len(f.p.Biomes) == 0 || f.p.RegionCell <= 0 {
		return cell{}, Biome{}
	}
	wx, wy := f.warp(f.compress(x, y))
	baseX, baseY := math.Floor(wx/f.p.RegionCell), math.Floor(wy/f.p.RegionCell)
	nearest, other := math.Inf(1), math.Inf(1)
	owner, neighbour := cell{}, ""
	ownerBiome := ""
	for dy := -2.0; dy <= 2; dy++ {
		for dx := -2.0; dx <= 2; dx++ {
			at := cell{int64(baseX + dx), int64(baseY + dy)}
			site, biome := f.site(at)
			distance := math.Sqrt((site.x-wx)*(site.x-wx) + (site.y-wy)*(site.y-wy))
			if distance < nearest {
				nearest, owner, ownerBiome = distance, at, biome
			}
		}
	}
	for dy := -2.0; dy <= 2; dy++ {
		for dx := -2.0; dx <= 2; dx++ {
			at := cell{int64(baseX + dx), int64(baseY + dy)}
			site, biome := f.site(at)
			if biome == ownerBiome {
				continue
			}
			distance := math.Sqrt((site.x-wx)*(site.x-wx) + (site.y-wy)*(site.y-wy))
			if distance < other {
				other, neighbour = distance, biome
			}
		}
	}
	blend := 1.0
	if width := f.p.BlendWidth * f.p.RegionCell; width > 0 && !math.IsInf(other, 1) {
		blend = clamp((other-nearest)/width, 0, 1)
	}
	return owner, Biome{ID: ownerBiome, Neighbour: neighbour, Blend: blend}
}

type point struct{ x, y float64 }

// site is one lattice site: its jittered position and the biome it owns. Both
// are drawn from the seed and the site's own coordinates alone, so the field is
// the same however it is sampled and in whatever order.
func (f *Field) site(at cell) (point, string) {
	ix, iy := uint32(int32(at.x)), uint32(int32(at.y))
	base := hash32(f.p.Seed^saltSite, ix, iy)
	jitter := clamp(f.p.RegionJitter, 0, 1)
	site := point{
		x: (float64(at.x) + 0.5 + (unit(hash32(base, ix, saltJitterX))-0.5)*jitter) * f.p.RegionCell,
		y: (float64(at.y) + 0.5 + (unit(hash32(base, iy, saltJitterY))-0.5)*jitter) * f.p.RegionCell,
	}
	index := int(unit(hash32(base, saltBiome, 0)) * float64(len(f.p.Biomes)))
	if index >= len(f.p.Biomes) {
		index = len(f.p.Biomes) - 1
	}
	return site, f.p.Biomes[index]
}

// compress pulls a position toward the reference radius by a fixed 3/4 power,
// which is written as two square roots because that is the only way Go and
// JavaScript are guaranteed to agree on it to the bit.
//
// The effect is the whole reason biome and grade are independent axes: a region
// becomes a swathe that narrows toward the hub and widens toward the rim rather
// than a blob at one radius, so every element is reachable from every band.
func (f *Field) compress(x, y float64) (float64, float64) {
	distance := math.Sqrt(x*x + y*y)
	reference := f.p.RadialReference
	if distance < 1 || reference <= 0 {
		return x, y
	}
	ratio := reference / distance
	scale := math.Sqrt(ratio * math.Sqrt(ratio)) // ratio ^ (3/4)
	return x * scale, y * scale
}

// warp bends the lattice with two octaves of value noise so region borders read
// as coastlines rather than as straight Voronoi seams.
func (f *Field) warp(x, y float64) (float64, float64) {
	if f.p.WarpCell <= 0 || f.p.WarpAmplitude == 0 {
		return x, y
	}
	u, v := x/f.p.WarpCell, y/f.p.WarpCell
	dx := f.noise(saltWarpX, u, v) + 0.5*f.noise(saltWarpX2, u*2, v*2)
	dy := f.noise(saltWarpY, u, v) + 0.5*f.noise(saltWarpY2, u*2, v*2)
	return x + f.p.WarpAmplitude*dx, y + f.p.WarpAmplitude*dy
}

// noise is ordinary bilinear value noise in [-1,1], smoothstepped so the warp
// has no lattice-aligned creases in it.
func (f *Field) noise(salt uint32, x, y float64) float64 {
	gx, gy := math.Floor(x), math.Floor(y)
	sx, sy := smooth(x-gx), smooth(y-gy)
	ix, iy := uint32(int32(gx)), uint32(int32(gy))
	seed := f.p.Seed ^ salt
	n00, n10 := unit(hash32(seed, ix, iy)), unit(hash32(seed, ix+1, iy))
	n01, n11 := unit(hash32(seed, ix, iy+1)), unit(hash32(seed, ix+1, iy+1))
	top := n00 + (n10-n00)*sx
	bottom := n01 + (n11-n01)*sx
	return (top+(bottom-top)*sy)*2 - 1
}

const (
	saltSite    uint32 = 0x2545f491
	saltJitterX uint32 = 0x51ed270b
	saltJitterY uint32 = 0x2f9277b5
	saltBiome   uint32 = 0x9e3779b1
	saltWarpX   uint32 = 0x1b56c4e9
	saltWarpY   uint32 = 0x7f4a7c15
	saltWarpX2  uint32 = 0x27d4eb2f
	saltWarpY2  uint32 = 0x165667b1
)

// hash32 is a three-input integer hash. Every step is 32-bit multiply, xor, and
// logical shift, which is exactly the set JavaScript reproduces bit for bit.
func hash32(a, b, c uint32) uint32 {
	h := a * 0x9e3779b1
	h ^= b + 0x85ebca6b + (h << 6) + (h >> 2)
	h ^= c + 0xc2b2ae35 + (h << 6) + (h >> 2)
	h ^= h >> 16
	h *= 0x7feb352d
	h ^= h >> 15
	h *= 0x846ca68b
	h ^= h >> 16
	return h
}

// unit maps a hash onto [0,1). The divisor is a power of two, so the division
// is exact and needs no rounding agreement between the two languages.
func unit(h uint32) float64 { return float64(h) / 4294967296 }

func smooth(t float64) float64 { return t * t * (3 - 2*t) }

func clamp(value, low, high float64) float64 {
	return math.Max(low, math.Min(high, value))
}

File: web/src/game/worldfield.ts
// The browser's copy of the world field. It is a line-for-line mirror of
// server/internal/worldfield/worldfield.go and must stay bit-identical to it:
// the renderer tints the ground by biome and the HUD names the region a player
// is standing in, and a client that disagreed with the server about where a
// border runs would be showing a different world from the one being simulated.
//
// Two rules keep that identity cheap, and both are why this file looks the way
// it does. Every hash is 32-bit integer arithmetic, reproduced here with
// Math.imul and the unsigned shift operators — a 64-bit hash would need BigInt.
// Every float operation is +, -, *, / or Math.sqrt, all of which IEEE-754
// specifies exactly; Math.pow is deliberately absent, because neither language
// is required to round it the same way.
import { biomes, materials, world } from "../tuning";

export interface Band {
  id: string; name: string; pvp: string; material_grade: string; shape: string; summary: string;
  tier: number; outer_radius: number;
}
export interface GradePoint { at: number; richness: number }
export interface GradeThreshold { id: string; tier: number; at: number }
export interface GradeCurve { points: GradePoint[]; thresholds: GradeThreshold[] }

export interface FieldParams {
  radius: number; seed: number; regionCell: number; regionJitter: number;
  radialReference: number; warpCell: number; warpAmplitude: number; blendWidth: number;
  biomeIDs: string[]; bands: Band[]; grades: GradeCurve;
}

/** What covers a position: its region's biome, the nearest different one, and
 *  how far from their shared border it sits — 0 on the border, 1 deep inside. */
export interface BiomeSample { id: string; neighbour: string; blend: number }
/** What the ground is worth: the grade it yields, and the convex richness the
 *  yield scales with inside that grade. */
export interface GradeSample { id: string; tier: number; richness: number }
export interface RegionSample { id: string; danger: Band; biome: BiomeSample; grade: GradeSample }

const SALT_SITE = 0x2545f491;
const SALT_JITTER_X = 0x51ed270b;
const SALT_JITTER_Y = 0x2f9277b5;
const SALT_BIOME = 0x9e3779b1;
const SALT_WARP_X = 0x1b56c4e9;
const SALT_WARP_Y = 0x7f4a7c15;
const SALT_WARP_X2 = 0x27d4eb2f;
const SALT_WARP_Y2 = 0x165667b1;

/** A three-input 32-bit hash. Every step is a 32-bit multiply, xor, or logical
 *  shift, which is exactly the set Go reproduces bit for bit. */
function hash32(a: number, b: number, c: number): number {
  let h = Math.imul(a, 0x9e3779b1) >>> 0;
  h = (h ^ ((b + 0x85ebca6b + ((h << 6) >>> 0) + (h >>> 2)) >>> 0)) >>> 0;
  h = (h ^ ((c + 0xc2b2ae35 + ((h << 6) >>> 0) + (h >>> 2)) >>> 0)) >>> 0;
  h = (h ^ (h >>> 16)) >>> 0;
  h = Math.imul(h, 0x7feb352d) >>> 0;
  h = (h ^ (h >>> 15)) >>> 0;
  h = Math.imul(h, 0x846ca68b) >>> 0;
  return (h ^ (h >>> 16)) >>> 0;
}

/** Maps a hash onto [0,1). The divisor is a power of two, so the division is
 *  exact and needs no rounding agreement between the two languages. */
function unit(h: number): number { return h / 4294967296; }
function index32(value: number): number { return (value | 0) >>> 0; }
function smooth(t: number): number { return t * t * (3 - 2 * t); }
function clamp(value: number, low: number, high: number): number {
  return Math.max(low, Math.min(high, value));
}

export class WorldField {
  private readonly p: FieldParams;

  constructor(params: FieldParams) {
    this.p = { ...params, biomeIDs: [...params.biomeIDs].sort() };
  }

  /** The band containing a position. Anything past the rim resolves to the
   *  outermost band rather than to nothing. */
  dangerAt(x: number, y: number): Band {
    const distance = Math.sqrt(x * x + y * y);
    return this.p.bands.find((band) => distance <= band.outer_radius) ?? this.p.bands[this.p.bands.length - 1]!;
  }

  protectedAt(x: number, y: number): boolean {
    const pvp = this.dangerAt(x, y).pvp;
    return pvp === "off" || pvp === "restricted";
  }

  safeAt(x: number, y: number): boolean { return this.dangerAt(x, y).pvp === "off"; }

  /** The convex reward curve, in [0,1]. It rises slowly across the Fringe and
   *  steeply toward the Deadlands, which is what makes the middle bands a route
   *  rather than the best farm. */
  richnessAt(x: number, y: number): number {
    const points = this.p.grades.points;
    if (points.length === 0 || this.p.radius <= 0) return 0;
    const at = Math.sqrt(x * x + y * y) / this.p.radius;
    if (at <= points[0]!.at) return points[0]!.richness;
    for (let index = 1; index < points.length; index++) {
      const previous = points[index - 1]!, current = points[index]!;
      if (at > current.at) continue;
      const span = current.at - previous.at;
      if (span <= 0) return current.richness;
      return previous.richness + (current.richness - previous.richness) * (at - previous.at) / span;
    }
    return points[points.length - 1]!.richness;
  }

  gradeAt(x: number, y: number): GradeSample {
    const richness = this.richnessAt(x, y);
    const grade: GradeSample = { id: "", tier: 0, richness };
    for (const threshold of this.p.grades.thresholds) {
      if (richness >= threshold.at) { grade.id = threshold.id; grade.tier = threshold.tier; }
    }
    return grade;
  }

  biomeAt(x: number, y: number): BiomeSample { return this.region(x, y).biome; }

  regionAt(x: number, y: number): RegionSample {
    const { cellX, cellY, biome } = this.region(x, y);
    return {
      id: `region-${cellX}:${cellY}`,
      danger: this.dangerAt(x, y),
      biome,
      grade: this.gradeAt(x, y),
    };
  }

  /**
   * Resolves the owning lattice site and the border blend around it.
   *
   * The search is five cells wide rather than three: a site may sit anywhere
   * inside its own cell, so the nearest site to a point at a cell corner can
   * still be two cells away on an axis, and a region border that flickers is
   * worse than one that costs twenty-five hashes.
   */
  private region(x: number, y: number): { cellX: number; cellY: number; biome: BiomeSample } {
    if (this.p.biomeIDs.length === 0 || this.p.regionCell <= 0) {
      return { cellX: 0, cellY: 0, biome: { id: "", neighbour: "", blend: 0 } };
    }
    const [cx, cy] = this.compress(x, y);
    const [wx, wy] = this.warp(cx, cy);
    const baseX = Math.floor(wx / this.p.regionCell), baseY = Math.floor(wy / this.p.regionCell);
    let nearest = Infinity, other = Infinity;
    let ownerX = 0, ownerY = 0, ownerBiome = "", neighbour = "";
    for (let dy = -2; dy <= 2; dy++) {
      for (let dx = -2; dx <= 2; dx++) {
        const [site, biome] = this.site(baseX + dx, baseY + dy);
        const distance = Math.sqrt((site[0] - wx) * (site[0] - wx) + (site[1] - wy) * (site[1] - wy));
        if (distance < nearest) { nearest = distance; ownerX = baseX + dx; ownerY = baseY + dy; ownerBiome = biome; }
      }
    }
    for (let dy = -2; dy <= 2; dy++) {
      for (let dx = -2; dx <= 2; dx++) {
        const [site, biome] = this.site(baseX + dx, baseY + dy);
        if (biome === ownerBiome) continue;
        const distance = Math.sqrt((site[0] - wx) * (site[0] - wx) + (site[1] - wy) * (site[1] - wy));
        if (distance < other) { other = distance; neighbour = biome; }
      }
    }
    let blend = 1;
    const width = this.p.blendWidth * this.p.regionCell;
    if (width > 0 && Number.isFinite(other)) blend = clamp((other - nearest) / width, 0, 1);
    return { cellX: ownerX, cellY: ownerY, biome: { id: ownerBiome, neighbour, blend } };
  }

  /** One lattice site: its jittered position and the biome it owns, both drawn
   *  from the seed and the site's own coordinates alone. */
  private site(cellX: number, cellY: number): [[number, number], string] {
    const ix = index32(cellX), iy = index32(cellY);
    const base = hash32((this.p.seed ^ SALT_SITE) >>> 0, ix, iy);
    const jitter = clamp(this.p.regionJitter, 0, 1);
    const site: [number, number] = [
      (cellX + 0.5 + (unit(hash32(base, ix, SALT_JITTER_X)) - 0.5) * jitter) * this.p.regionCell,
      (cellY + 0.5 + (unit(hash32(base, iy, SALT_JITTER_Y)) - 0.5) * jitter) * this.p.regionCell,
    ];
    let index = Math.trunc(unit(hash32(base, SALT_BIOME, 0)) * this.p.biomeIDs.length);
    if (index >= this.p.biomeIDs.length) index = this.p.biomeIDs.length - 1;
    return [site, this.p.biomeIDs[index]!];
  }

  /**
   * Pulls a position toward the reference radius by a fixed 3/4 power, written
   * as two square roots because that is the only way Go and JavaScript are
   * guaranteed to agree on it to the bit.
   *
   * The effect is why biome and grade are independent axes: a region becomes a
   * swathe that narrows toward the hub and widens toward the rim, so every
   * element is reachable from every danger band.
   */
  private compress(x: number, y: number): [number, number] {
    const distance = Math.sqrt(x * x + y * y);
    if (distance < 1 || this.p.radialReference <= 0) return [x, y];
    const ratio = this.p.radialReference / distance;
    const scale = Math.sqrt(ratio * Math.sqrt(ratio));
    return [x * scale, y * scale];
  }

  /** Bends the lattice with two octaves of value noise, so region borders read
   *  as coastlines rather than as straight Voronoi seams. */
  private warp(x: number, y: number): [number, number] {
    if (this.p.warpCell <= 0 || this.p.warpAmplitude === 0) return [x, y];
    const u = x / this.p.warpCell, v = y / this.p.warpCell;
    const dx = this.noise(SALT_WARP_X, u, v) + 0.5 * this.noise(SALT_WARP_X2, u * 2, v * 2);
    const dy = this.noise(SALT_WARP_Y, u, v) + 0.5 * this.noise(SALT_WARP_Y2, u * 2, v * 2);
    return [x + this.p.warpAmplitude * dx, y + this.p.warpAmplitude * dy];
  }

  /** Bilinear value noise in [-1,1], smoothstepped so the warp has no
   *  lattice-aligned creases in it. */
  private noise(salt: number, x: number, y: number): number {
    const gx = Math.floor(x), gy = Math.floor(y);
    const sx = smooth(x - gx), sy = smooth(y - gy);
    const ix = index32(gx), iy = index32(gy);
    const seed = (this.p.seed ^ salt) >>> 0;
    const n00 = unit(hash32(seed, ix, iy)), n10 = unit(hash32(seed, (ix + 1) >>> 0, iy));
    const n01 = unit(hash32(seed, ix, (iy + 1) >>> 0)), n11 = unit(hash32(seed, (ix + 1) >>> 0, (iy + 1) >>> 0));
    const top = n00 + (n10 - n00) * sx;
    const bottom = n01 + (n11 - n01) * sx;
    return (top + (bottom - top) * sy) * 2 - 1;
  }
}

/** Projects the shipped tables onto the field's inputs, mirroring
 *  `Tables.FieldParams` so both languages build the same field from the same
 *  rows rather than from two hand-kept copies. */
export function fieldParams(): FieldParams {
  const field = world.field;
  return {
    radius: world.radius,
    seed: field.seed >>> 0,
    regionCell: field.region_cell,
    regionJitter: field.region_jitter,
    radialReference: field.radial_reference,
    warpCell: field.warp_cell,
    warpAmplitude: field.warp_amplitude,
    blendWidth: field.blend_width,
    biomeIDs: Object.keys(biomes).sort(),
    bands: world.danger_bands,
    grades: {
      points: field.grade_curve.points.map(([at, richness]) => ({ at, richness })),
      thresholds: field.grade_curve.thresholds.map((threshold) => ({
        id: threshold.grade,
        tier: materials.grades[threshold.grade]?.tier ?? 0,
        at: threshold.at,
      })),
    },
  };
}

/** The shipped world field. One instance: it is stateless and pure. */
export const worldField = new WorldField(fieldParams());

/** A biome's ambient palette as renderer colours. An unknown biome — which is
 *  only reachable from a field with no biomes at all — falls back to a neutral
 *  slate rather than to black, so a broken table is legible instead of invisible. */
export function biomePalette(id: string): { ground: number; accent: number; haze: number } {
  const palette = biomes[id]?.palette;
  return {
    ground: hexColour(palette?.ground, 0x16233a),
    accent: hexColour(palette?.accent, 0x2c405e),
    haze: hexColour(palette?.haze, 0x7ee1bb),
  };
}

function hexColour(value: string | undefined, fallback: number): number {
  if (!value || value.length !== 7 || value[0] !== "#") return fallback;
  const parsed = Number.parseInt(value.slice(1), 16);
  return Number.isNaN(parsed) ? fallback : parsed;
}

/** Blends two packed colours, `t` of the way from `from` to `to`. */
export function mixColour(from: number, to: number, t: number): number {
  const clamped = clamp(t, 0, 1);
  const channel = (shift: number): number =>
    Math.round(((from >> shift) & 0xff) + (((to >> shift) & 0xff) - ((from >> shift) & 0xff)) * clamped);
  return (channel(16) << 16) | (channel(8) << 8) | channel(0);
}

/** The biome's display name, for the HUD readout. */
export function biomeName(id: string): string { return biomes[id]?.name ?? id; }

/** The material grade's display name. */
export function gradeName(id: string): string { return materials.grades[id]?.name ?? id; }

/**
 * What ground of a biome and grade tier can yield, mirroring
 * `Tables.MaterialsAt`: universal stock everywhere, the biome's own aligned
 * rows, and grade as a ceiling rather than an equality, so walking outward adds
 * options instead of trading them.
 */
export function materialsAt(biome: string, tier: number): string[] {
  const found: string[] = [];
  for (const [id, material] of Object.entries(materials.materials)) {
    const kind = materials.kinds[material.kind];
    if (!kind || kind.source === "craft") continue;
    const grade = materials.grades[material.grade];
    if (!grade || grade.tier > tier) continue;
    if (!kind.universal && material.biome !== biome) continue;
    found.push(id);
  }
  return found.sort();
}
